src/Language/DifferentialDatalog/Validate.hs

```diff
@@ -314,6 +319,93 @@ ruleCheckAggregate d rl idx = do
+                           $ "Transformer " ++ name trans ++ " expects a function that takes " ++ show (length hotArgs) ++ " arguments " ++
+                             " but function " ++ name f ++ " takes " ++ show (length funcArgs) ++ " arguments"
+                     mapM_ (\(farg, carg) -> check (argMut farg == argMut carg) (pos a) $
+                                             "Argument " ++ name farg ++ " of formal argument " ++ name decl ++ " of transformer " ++ name trans ++
+                                             " differs in mutability from argument " ++ name carg ++ " of function " ++ name f)
+                           $ zip hotArgs funcArgs
+                     return $ (zip (map typ hotArgs) (map typ funcArgs)) ++ [(hotType, funcType)]
+                 hot@HOTypeRelation{..} -> do
+                     rel <- checkRelation (pos a) d conc
+                     return [(hotType, relType rel)]
+                  ) $ zip (transInputs ++ transOutputs) (applyInputs ++ applyOutputs)
+    bindings <- unifyTypes d (pos a) ("in transformer application " ++ show a) $ concat types
+    mapM_ (\ta -> case M.lookup ta bindings of
+                       Nothing -> err (pos a) $ "Unable to bind type argument '" ++ ta ++
+                                                " to a concrete type in transformer application " ++ show a
+                       Just t  -> return ())
+          $ transformerTypeVars trans
+    mapM_ (\o -> check (relRole (getRelation d o) /= RelInput) (pos a)
+                 $ "Transformer output cannot be bound to input relation " ++ o
+          ) applyOutputs
+    -- Things will break if a relation is assigned by an 'Apply' in the top scope and then occurs inside a
+    -- recursive fragment.  Keep things simple by disallowing this.
+    -- If this proves an important limitation, it can be dropped and replaced with a program
+    -- transformation that introduces a new relation for each 'Apply' output and a rule that
+    -- concatenates it to the original output relation.  But for now it seems like a useful
+    -- restriction
+    mapM_ (\o -> check (null $ relRules d o) (pos a)
+                       $ "Output of a transformer application may not occur in the head of a rule, but relation " ++ o ++
+                          " occurs in the following rules\n" ++ (intercalate "\n" $ map show $ relRules d o))
+          applyOutputs
+    -- Likewise, to relax this, modify 'compileApplyNode' to concatenate transformer output to
+    -- existing relation if it exists.
+    mapM_ (\o -> check (length (relApplys d o) == 1) (pos a)
+                       $ "Relation " ++ o ++ " occurs as output of multiple transformer applications")
+          applyOutputs
+
+hotypeValidate :: (MonadError String me) => DatalogProgram -> HOType -> me ()
+hotypeValidate d HOTypeFunction{..} = do
+    -- FIXME: hacky way to validate function type by converting it into a function.
```

Comment:
why is it hacky?

Reply:
A clean solution is to change the function type to basically be a named variable of higher-order type.